general/echo/kmdf/exe/src/main.rs

```diff
@@ -0,0 +1,489 @@
+                }
+
+                if remaining_requests_to_send == 0 {
+                    continue;
+                }
+
+                remaining_requests_to_send -= 1;
+            }
+
+            unsafe {
+                r = ReadFile(
+                    h_device,
+                    (buf.as_mut_ptr().offset(i * (BUFFER_SIZE as isize))) as _,
+                    BUFFER_SIZE as u32,
+                    std::ptr::null_mut(),
+                    completed_ov_ptr,
+                );
+
+                if r == 0 {
+                    let error = GetLastError();
+                    if error != ERROR_IO_PENDING {
+                        return Err(String::from(format!("{}th Read failed {}", i, error)).into());
+                    }
+                }
+            }
+        } else {
+            println!(
+                "Number of bytes written by request number {} is {}",
+                i, number_of_bytes_transferred
+            );
+
+            if globals.g_limited_loops {
+                remaining_requests_to_receive -= 1;
+                if remaining_requests_to_receive == 0 {
+                    break;
+                }
+
+                if remaining_requests_to_send == 0 {
+                    continue;
+                }
+
+                remaining_requests_to_send -= 1;
+            }
+
+            unsafe {
+                r = WriteFile(
+                    h_device,
+                    (buf.as_mut_ptr().offset(i * (BUFFER_SIZE as isize))) as _,
+                    BUFFER_SIZE as u32,
+                    std::ptr::null_mut(),
+                    completed_ov_ptr,
+                );
+
+                if r == 0 {
+                    let error = GetLastError();
+                    if error != ERROR_IO_PENDING {
+                        return Err(String::from(format!("{}th write failed {}", i, error)).into());
+                    }
+                }
+            }
+        }
+    }
+
+    unsafe {
+        CloseHandle(h_completion_port);
+
+        CloseHandle(h_device);
+    }
+
+    Ok(())
+}
+
+fn get_device_path(interface_guid: &Uuid) -> Result<(), Box<dyn Error>> {
+    let mut guid = windows_sys::core::GUID {
+        data1: 0,
+        data2: 0,
+        data3: 0,
+        data4: [0, 0, 0, 0, 0, 0, 0, 0],
+    };
+    let guid_data4: &[u8; 8];
+    let mut device_interface_list_length: u32 = 0;
+    let mut config_ret;
+
+    (guid.data1, guid.data2, guid.data3, guid_data4) = interface_guid.as_fields();
+    guid.data4 = *guid_data4;
+
+    unsafe {
+        config_ret = windows_sys::Win32::Devices::DeviceAndDriverInstallation::CM_Get_Device_Interface_List_SizeW(&mut device_interface_list_length,
+            &guid, std::ptr::null(), windows_sys::Win32::Devices::DeviceAndDriverInstallation::CM_GET_DEVICE_INTERFACE_LIST_PRESENT);
+    }
+
+    if config_ret != windows_sys::Win32::Devices::DeviceAndDriverInstallation::CR_SUCCESS {
+        return Err(String::from(format!(
```

Comment:
`format!` creates an owned `String` so `String::from` is redundant

Reply:
All of these cleaned up